### ISSUE TITLE
`MimirIngesterHasUnshippedBlocks` can fire even though block was uploaded

### ISSUE THREAD

**Issue by pstibrany** [grafana/mimir#1997]
#### Describe the bug

`MimirIngesterHasUnshippedBlocks` checks oldest not-yet uploaded block (`cortex_ingester_oldest_unshipped_block_timestamp_seconds`) was generated over 1h ago. But due to a bug in ingester, `cortex_ingester_oldest_unshipped_block_timestamp_seconds` may not be updated even though block was actually uploaded.

In observed case, here is what happened:

- Ingester generated block, and shipper tried to upload it
- Upload of meta.json failed due to 503 error, but meta.json was actually uploaded
- On next shipper iteration, shipper detected that meta.json file already exists, and didn't upload it. In this case shipper considers block to be "uploaded", but it doesn't update `uploaded` counter: https://github.com/grafana/mimir/blob/2c2c65d69bdf466ea66bfa7d4e87704ec9a7f426/pkg/ingester/shipper.go#L157-L164
- Ingester only updates cached version of uploaded blocks if shipper reports any uploaded block. But since it didn't (see previous step), ingester didn't update cached version of shipped blocks: https://github.com/grafana/mimir/blob/0772991cf394d52dff17ccdcb0c52a4652abe706/pkg/ingester/ingester.go#L1797-L1806
- Ingester's `cortex_ingester_oldest_unshipped_block_timestamp_seconds` is based on cached version of shipped blocks.

This is why Mimir may report incorrect `cortex_ingester_oldest_unshipped_block_timestamp_seconds` for another 2 hours, which may cause `MimirIngesterHasUnshippedBlocks` to fire.

#### Expected behavior

No `MimirIngesterHasUnshippedBlocks` if block was actually uploaded.

**Comment by williamzelesny** [comment]
curious if this is the right way to go about mitigating this situation

**Comment by jhesketh** [comment]
Please excuse my ignorance here, still learning things.

I'm curious if it is sufficient to simply check if the meta file exists in the bucket. Do we understand why it "failed" to upload in the first case? Perhaps it was a partial upload or corrupted etc. Just knowing whether it exists or not may not tell us if we need to upload it again.

Additionally I'm confused about the case where blocks are uploaded by a different ingester. These are skipped over by 
https://github.com/grafana/mimir/blob/e11ac855ba074308735271208931c20bc2ce0445/pkg/ingester/shipper.go#L147-L155
which are then marked as uploaded, but because the `uploaded` counter is never incremented if an ingester never does any other upload, then its cache won't get updated and will surely show `MimirIngesterHasUnshippedBlocks`?

In other words, do we not need to increment `uploaded` in every case even if this particular ingester hasn't performed the uploaded?

(Sorry if that doesn't make sense!)

**Comment by pstibrany** [comment]
> Please excuse my ignorance here, still learning things.
> 
> I'm curious if it is sufficient to simply check if the meta file exists in the bucket. Do we understand why it "failed" to upload in the first case? Perhaps it was a partial upload or corrupted etc. Just knowing whether it exists or not may not tell us if we need to upload it again.

That's a good point. We don't why it failed, only that it failed. From what I can see neither GCS or S3 would store half-uploaded or corrupted file, so checking for file existence is enough for them. But it may not be true for some other block storages, and it may be worth downloading and parsing json file to make sure it's complete.

> Additionally I'm confused about the case where blocks are uploaded by a different ingester. These are skipped over by
> 
> https://github.com/grafana/mimir/blob/e11ac855ba074308735271208931c20bc2ce0445/pkg/ingester/shipper.go#L147-L155

Blocks cannot be uploaded by different ingester. Each ingester creates blocks with unique ID (ULID), and each ingester can only upload its own blocks.

> which are then marked as uploaded, but because the `uploaded` counter is never incremented if an ingester never does any other upload, then its cache won't get updated and will surely show `MimirIngesterHasUnshippedBlocks`?
> In other words, do we not need to increment `uploaded` in every case even if this particular ingester hasn't performed the uploaded?

Yes, updating `uploaded` in case when `meta.json` exists is one way of solving the problem, although I think it would lead to slightly confusing log line as block was actually uploaded in previous iteration.

**Comment by jhesketh** [comment]
> That's a good point. We don't why it failed, only that it failed. From what I can see neither GCS or S3 would store half-uploaded or corrupted file, so checking for file existence is enough for them. But it may not be true for some other block storages, and it may be worth downloading and parsing json file to make sure it's complete.

I had thought about that but figured the overhead cost would be too much. At least in the case of doing it whenever the block already exists, but it could be affordable only during failures (such as during the check in #2059).

> Blocks cannot be uploaded by different ingester. Each ingester creates blocks with unique ID (ULID), and each ingester can only upload its own blocks.

Ah, okay, that's the part I was misunderstanding. Thanks!

In this case every block wants to be uploaded, so #2059 makes sense to me (with the possible expansion of checking the contents). An alternative which may be safer against object stores where we aren't confident they are storing the whole file could be to always reupload any failures (or delete any files that exist after failure).

**Comment by pstibrany** [comment]
> > That's a good point. We don't why it failed, only that it failed. From what I can see neither GCS or S3 would store half-uploaded or corrupted file, so checking for file existence is enough for them. But it may not be true for some other block storages, and it may be worth downloading and parsing json file to make sure it's complete.
> 
> I had thought about that but figured the overhead cost would be too much. At least in the case of doing it whenever the block already exists, but it could be affordable only during failures (such as during the check in #2059).

I don't think there would be any extra overhead. Ingester produces one block per tenant, every two hours. Right now it issues one extra "Exists" call before uploading the block. Normally block wasn't yet uploaded and we get "does not exist". If we changed it to "Get", number of operations stays the same, and in case that meta.json file actually does exist (because it was uploaded before, but shipper didn't record it in its local json file), we can check if it is correct.

> In this case every block wants to be uploaded, so #2059 makes sense to me (with the possible expansion of checking the contents). An alternative which may be safer against object stores where we aren't confident they are storing the whole file could be to always reupload any failures (or delete any files that exist after failure).

"could be to always reupload any failures" -- that means removing check for `meta.json` file before we do upload. That would work too, but if meta.json is there it means all other block files are there too (we upload meta.json as last), so reuploading entire block is a waste of time and bandwidth. However it is harmless.

**Comment by dimitarvdimitrov** [comment]
the original issue is that the alert fires even though the block upload completed. At every ship interval the shipper updates the local `thanos.shipper.json` file with the uploaded blocks: block that completed uploading are added to to the `uploaded` field there ([code](https://github.com/grafana/mimir/blob/8bff6d023ae1581f61061e4c6fef66cc604aa4c7/pkg/ingester/shipper.go#L164)) and ones that are discovered as already uploaded by the `Exists` check are also added to the file ([code](https://github.com/grafana/mimir/blob/8bff6d023ae1581f61061e4c6fef66cc604aa4c7/pkg/ingester/shipper.go#L153)). The problem is that we call [`updateCachedShippedBlocks`](https://github.com/grafana/mimir/blob/0772991cf394d52dff17ccdcb0c52a4652abe706/pkg/ingester/user_tsdb.go#L252) only when there is at [least one successfully uploaded block](https://github.com/grafana/mimir/blob/4679c7549ef2e80c4d6a79a1c93175a0259ca939/pkg/ingester/ingester.go#L1817-L1821). 

The cache update seems cheap - it's only reading a local file. One way to solve this is to just update the local shipper file on every sync. 

There's this comment which assumes we don't have to update the file

https://github.com/grafana/mimir/blob/4679c7549ef2e80c4d6a79a1c93175a0259ca939/pkg/ingester/ingester.go#L1814-L1817

but I think this comment doesn't account for _this_ issue

**Comment by pstibrany** [comment]
> The cache update seems cheap - it's only reading a local file. One way to solve this is to just update the local shipper file on every sync.

Sync (shipping) happens every minute by default. In a cluster with thousands of tenants reloading the file every minute seems like a bad idea.

**Comment by pstibrany** [comment]
> the file is already being read on every Sync in the shipper

Ha, good point :) Still, if we already know when file is modified (`Sync` can report that), why do another check?
